pkg/conn/basedb.go

```diff
@@ -76,11 +79,14 @@ type BaseDB struct {
 // NewBaseDB returns *BaseDB object
 func NewBaseDB(db *sql.DB) *BaseDB {
 	conns := make([]*BaseConn, 0)
-	return &BaseDB{db, conns, &retry.FiniteRetryStrategy{}}
+	lock := new(sync.Mutex)
+	return &BaseDB{db, lock, conns, &retry.FiniteRetryStrategy{}}
 }
 
 // GetBaseConn retrieves *BaseConn which has own retryStrategy
 func (d *BaseDB) GetBaseConn(ctx context.Context) (*BaseConn, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
```

Comment:
can we move this block before `d.conns = append(d.conns, baseConn)`?

Reply:
sure, [moved](https://github.com/pingcap/dm/pull/266/files#diff-184b4272aa0f082766a0d470548f0644R95)